include/vinecopulib/vinecop/implementation/class.ipp

```diff
@@ -657,46 +732,72 @@ Vinecop::pdf(const Eigen::MatrixXd& u, const size_t num_threads) const
     needed_hfunc1 = vine_struct_.get_needed_hfunc1();
     needed_hfunc2 = vine_struct_.get_needed_hfunc2();
   }
+  auto disc_cols = tools_select::get_disc_cols(var_types_);
 
   // initial value must be 1.0 for multiplication
-  Eigen::VectorXd vine_density = Eigen::VectorXd::Constant(u.rows(), 1.0);
+  Eigen::VectorXd pdf = Eigen::VectorXd::Constant(u.rows(), 1.0);
 
   auto do_batch = [&](const tools_batch::Batch& b) {
-    // temporary storage objects for h-functions
+    // temporary storage objects (all data must be in (0, 1))
     Eigen::MatrixXd hfunc1(b.size, d);
-    Eigen::MatrixXd hfunc2(b.size, d);
-    Eigen::MatrixXd u_e(b.size, 2);
+    hfunc1.setZero();
+    Eigen::MatrixXd hfunc2 = hfunc1;
+    Eigen::MatrixXd hfunc1_sub = hfunc1;
+    Eigen::MatrixXd hfunc2_sub = hfunc1;
+    Eigen::MatrixXd u_e(b.size, 4);
+    Eigen::MatrixXd u_sub(b.size, 4);
 
     // fill first row of hfunc2 matrix with evaluation points;
     // points have to be reordered to correspond to natural order
-    for (size_t j = 0; j < d; ++j)
+
+    for (size_t j = 0; j < d; ++j) {
       hfunc2.col(j) = u.block(b.begin, order[j] - 1, b.size, 1);
+      if (var_types_[order[j] - 1] == "d") {
+        hfunc2_sub.col(j) =
+          u.block(b.begin, d_ + disc_cols[order[j] - 1], b.size, 1);
+      } else {
+        hfunc2_sub.col(j) = u.block(b.begin, order[j] - 1, b.size, 1);
+      }
+    }
 
     for (size_t tree = 0; tree < trunc_lvl; ++tree) {
       tools_interface::check_user_interrupt(n * d > 1e5);
       for (size_t edge = 0; edge < d - tree - 1; ++edge) {
         tools_interface::check_user_interrupt(edge % 100 == 0);
         // extract evaluation point from hfunction matrices (have been
         // computed in previous tree level)
-        size_t m = min_array(tree, edge);
         u_e.col(0) = hfunc2.col(edge);
+        u_e.col(2) = hfunc2_sub.col(edge);
```

Comment:
So we do this even when both variables are continuous?

Reply:
Yeah, we could always do a check if both variables are continuous, but that makes the code a lot uglier. I would assume that copying the data has has almost impact on the performance, but we can measure.